server/src/main/java/org/elasticsearch/indices/recovery/plan/SnapshotsRecoveryPlannerService.java

```diff
@@ -0,0 +1,179 @@
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
+import static org.elasticsearch.common.util.CollectionUtils.concatLists;
+
+public class SnapshotsRecoveryPlannerService implements RecoveryPlannerService {
+    private final Logger logger = LogManager.getLogger(SnapshotsRecoveryPlannerService.class);
+
+    private final ShardSnapshotsService shardSnapshotsService;
+    private final boolean snapshotRecoveriesEnabled;
+    private final String repository;
+
+    public SnapshotsRecoveryPlannerService(ShardSnapshotsService shardSnapshotsService,
+                                           boolean snapshotRecoveriesEnabled,
+                                           @Nullable String repository) {
+        this.shardSnapshotsService = shardSnapshotsService;
+        this.snapshotRecoveriesEnabled = snapshotRecoveriesEnabled;
+        this.repository = repository;
+    }
+
+    public void computeRecoveryPlan(ShardId shardId,
+                                    String shardIdentifier,
+                                    Store.MetadataSnapshot sourceMetadata,
+                                    Store.MetadataSnapshot targetMetadata,
+                                    long startingSeqNo,
+                                    int translogOps,
+                                    ActionListener<ShardRecoveryPlan> listener) {
+        fetchAvailableSnapshotsIgnoringErrors(shardId, availableSnapshots ->
+            ActionListener.completeWith(listener, () ->
+                computeRecoveryPlanWithSnapshots(shardIdentifier,
+                    sourceMetadata,
+                    targetMetadata,
+                    startingSeqNo,
+                    translogOps,
+                    availableSnapshots
+                )
+            )
+        );
+    }
+
+    private ShardRecoveryPlan computeRecoveryPlanWithSnapshots(String shardIdentifier,
+                                                               Store.MetadataSnapshot sourceMetadata,
+                                                               Store.MetadataSnapshot targetMetadata,
+                                                               long startingSeqNo,
+                                                               int translogOps,
+                                                               List<ShardSnapshot> availableSnapshots) {
+        Store.RecoveryDiff sourceTargetDiff = sourceMetadata.recoveryDiff(targetMetadata);
+        List<StoreFileMetadata> filesMissingInTarget = concatLists(sourceTargetDiff.missing, sourceTargetDiff.different);
+
+        Store.MetadataSnapshot filesToRecoverFromSourceSnapshot = toMetadataSnapshot(filesMissingInTarget);
+
+        ShardRecoveryPlan plan = null;
+        int filesToRecoverFromSnapshot = 0;
+        for (ShardSnapshot snapshot : availableSnapshots) {
```

Comment:
If we think this might be confusing when we use multiple repositories, we might want to restrict this optimization to use only 1 repository?

Reply:
I agree to only use 1 repository, at least in the first version of this (and I would think it is good enough in general). I would be hesitant to consult all repositories, since some of them may be more expensive to access if located in a different region for instance.